rs/crypto/ecdsa_secp256k1/src/lib.rs

```diff
@@ -436,17 +438,42 @@ impl PrivateKey {
         }
 
         use k256::ecdsa::{signature::hazmat::PrehashSigner, Signature};
-        let sig: Signature = self
-            .key
-            .sign_prehash(digest)
-            .expect("Failed to sign digest");
+        let ecdsa = k256::ecdsa::SigningKey::from(&self.key);
+        let sig: Signature = ecdsa.sign_prehash(digest).expect("Failed to sign digest");
         Some(sig.to_bytes().into())
     }
 
+    /// Sign a message with BIP340 Schnorr
+    pub fn sign_bip340<R: Rng + CryptoRng>(&self, message: &[u8; 32], rng: &mut R) -> [u8; 64] {
+        let need_flip = self.public_key().serialize_sec1(true)[0] == 0x03;
+
+        let bip340 = if need_flip {
+            let ns = self.key.to_nonzero_scalar().negate();
+            let nz_ns =
+                k256::NonZeroScalar::new(ns).expect("Negation of non-zero is always non-zero");
+            k256::schnorr::SigningKey::from(nz_ns)
+        } else {
+            k256::schnorr::SigningKey::from(&self.key)
+        };
+
+        loop {
+            /*
+             * The only way this function can fail is the (cryptographically unlikely)
+             * situation where k or s of zero is generated. If this occurs, simply retry
+             * with a new aux_rand
+             */
+            let aux_rand = rng.gen::<[u8; 32]>();
+            if let Ok(sig) = bip340.sign_prehash_with_aux_rand(message, &aux_rand) {
```

Comment:
I am a bit confused why we use a `sign_prehash` function here, since we don't seem to be hashing the message. Is this the reason we only allow 32 bytes messages as inputs? For the bitcoin use cases this may be sufficient, but diverges with what the IC allows to sign. If we use this in state machine tests, e.g., we may have failures in tests that would succeed on the IC.

Reply:
This is just what `k256 0.13` calls this. In master it's called `sign_raw` and allows arbitrary inputs.